apps/crn-server/src/data-providers/contentful/analytics.data-provider.ts

```diff
@@ -24,6 +25,87 @@ export class AnalyticsContentfulDataProvider implements AnalyticsDataProvider {
+          teamMembershipItem?.linkedFrom?.usersCollection?.items
+            .flatMap(flattenInterestGroupLeaders)
+            .filter(
+              (item) => item.interestGroupActive && item.userIsAlumni === false,
+            )
+            .map((item) => item.interestGroupId) || [],
+      ) || [];
+
+    const getCurrentInterestGroupIdsFromInterestGroupCollection = (
+      team: Team,
+    ): string[] =>
+      team.linkedFrom?.interestGroupsCollection?.items
+        .filter(
+          (interestGroup): interestGroup is InterestGroup =>
+            (interestGroup && !!interestGroup.active) || false,
+        )
+        .flatMap((interestGroup) => interestGroup.sys.id) || [];
+
+    const getPreviousInterestGroupIdsFromTeamLeaders = (team: Team): string[] =>
+      team.linkedFrom?.teamMembershipCollection?.items.flatMap(
+        (teamMembershipItem) =>
+          teamMembershipItem?.linkedFrom?.usersCollection?.items
+            .flatMap(flattenInterestGroupLeaders)
+            .filter(
+              (item) =>
+                team.inactiveSince ||
+                item.userIsAlumni ||
+                !item.interestGroupActive,
+            )
+            .map((item) => item.interestGroupId) || [],
+      ) || [];
+
+    const getPreviousInterestGroupIdsFromInterestGroupCollection = (
+      team: Team,
+    ): string[] =>
+      team.linkedFrom?.interestGroupsCollection?.items
+        .filter(
+          (interestGroup): interestGroup is InterestGroup =>
+            interestGroup && (team.inactiveSince || !interestGroup?.active),
+        )
+        .flatMap((interestGroup) => interestGroup.sys.id) || [];
+
+    const getCurrentWorkingGroupIdsFromTeamMembers = (team: Team): string[] =>
+      team.linkedFrom?.teamMembershipCollection?.items.flatMap(
+        (teamMembershipItem) =>
+          teamMembershipItem?.linkedFrom?.usersCollection?.items
+            .flatMap(flattenWorkingGroupMember)
+            .filter((item) => !item.userIsAlumni && !item.workingGroupComplete)
+            .map((item) => item.workingGroupId) || [],
+      ) || [];
+
+    const getCurrentWorkingGroupIdsFromTeamLeaders = (team: Team): string[] =>
+      team.linkedFrom?.teamMembershipCollection?.items.flatMap(
+        (teamMembershipItem) =>
+          teamMembershipItem?.linkedFrom?.usersCollection?.items
+            .flatMap(flattenWorkingGroupLeaders)
+            .filter((item) => !item.userIsAlumni && !item.workingGroupComplete)
+            .map((item) => item.workingGroupId) || [],
+      ) || [];
+
+    const getPreviousWorkingGroupIdsFromTeamLeaders = (team: Team): string[] =>
+      team.linkedFrom?.teamMembershipCollection?.items.flatMap(
+        (teamMembershipItem) =>
+          teamMembershipItem?.linkedFrom?.usersCollection?.items
+            .flatMap(flattenWorkingGroupLeaders)
+            .filter((item) => item.userIsAlumni || item.workingGroupComplete)
+            .map((item) => item.workingGroupId) || [],
+      ) || [];
+
```

Comment:
What if we use a reduce to get the current and previous in one loop something like 


```
type EntityIdsByStatus = { current: string[]; previous: string[] };
const emptyEntityIdsByStatus: EntityIdsByStatus = {
  current: [],
  previous: [],
};

const getWorkingGroupIdsLeadership = (team: Team): EntityIdsByStatus =>
  team.linkedFrom?.teamMembershipCollection?.items.reduce(
    ({ current, previous }: EntityIdsByStatus, teamMembershipItem) => {
      teamMembershipItem?.linkedFrom?.usersCollection?.items
        .flatMap(flattenWorkingGroupLeaders)
        .forEach((item) => {
          if (
            item.userIsAlumni ||
            item.workingGroupComplete ||
            team.inactiveSince
          ) {
            previous.push(item.workingGroupId);
          } else {
            current.push(item.workingGroupId);
          }
        });
      return { current, previous };
    },
    emptyEntityIdsByStatus,
  ) || emptyEntityIdsByStatus;

(...)

    return {
      total: teamsCollection?.total || 0,
      items:
        teamsCollection?.items
          .filter((team): team is Team => team !== null)
          .map((team) => {
            const workingGroupIdsLeadership =
              getWorkingGroupIdsLeadership(team);

            return {
              id: team.sys.id,
              (...)
              workingGroupLeadershipRoleCount: team.inactiveSince
                ? 0
                : getUniqueIdCount(workingGroupIdsLeadership.current),
              workingGroupPreviousLeadershipRoleCount: getUniqueIdCount(
                workingGroupIdsLeadership.previous,
              ),
}
```

This way we go through this loop once instead of 3 times. I was wondering if we could do the same for working group members and interest group leaders as well. But it's just a suggestion.